ts/haskell/mod.ts

```diff
@@ -1,35 +1,72 @@
-import { packageToEnvironment, shell } from "../environment.ts";
+import { packageToEnvironment } from "../environment.ts";
 import { mkPackage, Package } from "../package.ts";
 import { mkProject, Project } from "../project.ts";
+import { Executable } from "../executable.ts";
 import { nixSource } from "../internal/utils.ts";
-import { nixRaw, nixStrLit } from "../nix.ts";
+import { nixRaw } from "../nix.ts";
+
+/**
+ * Project components returned by `mkHaskellProject`
+ */
+export type HaskellAddenda = {
+  /**
+   * A package building the entire Haskell project
+   */
+  pkg: Package;
+  /**
+   * Make an executable from the cabal file available to garn.
+   */
+  addCabalExecutable: <T extends Project & HaskellAddenda, Name extends string>(
+    this: T,
+    executableName: Name,
+  ) => Omit<T, Name> & { [n in Name]: Executable };
+};
+
+type Execs<Exe extends string[]> = {
+  [e in Exe[number]]: Executable;
+};
 
 /**
  * Creates a haskell-based garn Project.
+ *
+ * This will create a Project with
+ *   - a `pkg` key that builds the entire haskell project (i.e., `cabal build`);
+ *   - a default shell that provides cabal as well as all Haskell dependencies;
+ *   - one garn executable for each cabal executable in `executables`.
+ *
+ * @param description - A short description of the project
+ * @param compiler - The compiler version (e.g.: "ghc94")
+ * @param executables - The name of the executables in the cabal file
+ * @param src - The source directory
  */
-export function mkHaskellProject(args: {
+export function mkHaskellProject<const Executables extends string[]>(args: {
   description: string;
-  executable: string;
   compiler: string;
+  executables?: Executables;
   src: string;
-}): Project & { pkg: Package } {
+}): Project & HaskellAddenda & Execs<Executables> {
```

Comment:
Should we omit `Executables[number]` from `Project & HaskellAddenda`?

Reply:
Not sure I understand

Reply:
Yeah I didn't phrase that well. I meant something like this:

```suggestion
}): Omit<Project & HaskellAddenda, Executables[number]> & Execs<Executables> {
```

This way if someone has a Haskell executable named `check` for example, the `check` property on the `Project` is correctly typed as the Haskell executable instead of the method that we provide in `ProjectHelpers`.

Reply:
Just doing that results in an error in the test:

```
error: TS2349 [ERROR]: This expression is not callable.
  Type 'Executable' has no call signatures.
        .addCabalExecutable("foo");
         ~~~~~~~~~~~~~~~~~~
    at file:///home/jkarni/garnix/garn/ts/haskell/mod.test.ts:32:10
```

By making the `executables` argument readonly, I can get it to work this way when it's explicitly provided, but not when it's omitted.

In general, I do worry this sacrifices type inference by default assuming projects are projects....

Reply:
That said, your suggestion is really nice in actually giving a type error when I add an executable named `addCabalExecutable`...